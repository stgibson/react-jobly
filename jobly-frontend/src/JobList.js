import React, { useState } from "react";
import Container from "react-bootstrap/Container";
import Form from "react-bootstrap/Form";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";
import Button from "react-bootstrap/Button";
import JobCard from "./JobCard";

/**
 * Component for listing jobs with a search box for filtering jobs
 * @param
 * {Object{Object{Array[Object{number|string}]}|function}}
 * param0 
 * @returns JSX code for rendering job list and search box
 */
const JobList = ({ jobs, findAllJobs, apply }) => {
  const [filter, setFilter] = useState("");

  /**
   * Updates filter when user types in searchbox
   * @param {Object} evt 
   */
  const handleChange = evt => {
    const { value } = evt.target;
    setFilter(value);
  };

  /**
   * Stops reloading page and gets filtered jobs
   * @param {Object} evt 
   */
  const handleSubmit = evt => {
    evt.preventDefault();
    if (!filter) {
      findAllJobs();
    }
    else {
      findAllJobs(filter);
    }
  };

  // Learned how to use react-bootstrap forms at https://react-bootstrap.github.io/components/forms/
  // Learned how to use react-bootstrap container at https://react-bootstrap.github.io/layout/grid/
  // Learned how to change padding at https://mdbootstrap.com/docs/react/utilities/spacing/
  return (
    <Container>
      <Form onSubmit={ handleSubmit }>
        <Form.Group as={ Row } controlId="formSearchBox">
          <Col xs={ 10 } className="p-0">
            <Form.Control type="text" placeholder="Enter search term..." onChange={ handleChange } value={ filter } />
          </Col>
          <Col xs={ 2 }>
            <Button variant="primary" type="submit">Submit</Button>
          </Col>
        </Form.Group>
      </Form>
      <Row>
        <Col xs={ 12 } className="p-0">
          {
            jobs.map(job => (
              <JobCard
                key={ job.id }
                job={ job }
                apply={ apply }
              />
            ))
          }
        </Col>
      </Row>
    </Container>
  );
};

export default JobList;