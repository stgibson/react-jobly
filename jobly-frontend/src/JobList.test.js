import React from "react";
import { MemoryRouter } from "react-router-dom";
import { render, fireEvent } from "@testing-library/react";
import JobList from "./JobList";

const jobs = [
  {
    id: 1,
    title: "Test 1",
    salary: 100,
    equity: "0.1",
    companyHandle: "test-company-1",
    companyName: "Test Company 1"
  },
  {
    id: 2,
    title: "Test 2",
    salary: 200,
    equity: "0.2",
    companyHandle: "test-company-1",
    companyName: "Test Company 1"
  },
  {
    id: 3,
    title: "Test 3",
    salary: 300,
    equity: "0.3",
    companyHandle: "test-company-2",
    companyName: "Test Company 2"
  }
];

const findAllJobs = jest.fn(() => null); // Learned how to mock functions at https://jestjs.io/docs/mock-functions

it("renders without crashing", () => {
  render(<MemoryRouter><JobList jobs={ jobs } /></MemoryRouter>);
});

it("matches snapshot", () => {
  const { asFragment } = render(
    <MemoryRouter><JobList jobs={ jobs } /></MemoryRouter>
  );
  
  expect(asFragment()).toMatchSnapshot();
});

it("shows list of jobs", () => {
  const { queryByTestId } = render(
    <MemoryRouter><JobList jobs={ jobs } /></MemoryRouter>
  );

  for (let job of jobs) {
    expect(queryByTestId(job.id)).toBeInTheDocument();
  }
});

it("updates searchbox when user types", () => {
  const { queryByPlaceholderText, queryByDisplayValue } = render(
    <MemoryRouter><JobList jobs={ jobs } /></MemoryRouter>
  );

  const searchBox = queryByPlaceholderText("Enter search term...");
  
  fireEvent.change(searchBox, { target: { value: "1" } });

  expect(queryByDisplayValue("1")).toBeInTheDocument();
});

it("calls findAllJobs with no filter when searchbox is empty", () => {
  const { queryByText } = render(
    <MemoryRouter>
      <JobList jobs={ jobs } findAllJobs={ findAllJobs } />
    </MemoryRouter>
  );

  const submitButton = queryByText("Submit");
  
  fireEvent.click(submitButton);

  expect(findAllJobs.mock.calls.length).toEqual(1);
  expect(findAllJobs.mock.calls[0].length).toEqual(0);
});

it("calls findAllJobs with filter when searchbox is empty", () => {
  const { queryByPlaceholderText, queryByText } = render(
    <MemoryRouter>
      <JobList jobs={ jobs } findAllJobs={ findAllJobs } />
    </MemoryRouter>
  );

  const searchBox = queryByPlaceholderText("Enter search term...");
  
  fireEvent.change(searchBox, { target: { value: "1" } });

  const submitButton = queryByText("Submit");
  
  fireEvent.click(submitButton);

  expect(findAllJobs.mock.calls.length).toEqual(1);
  expect(findAllJobs.mock.calls[0][0]).toEqual("1");
});