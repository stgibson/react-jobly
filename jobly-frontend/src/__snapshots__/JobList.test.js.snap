// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <form
      class=""
    >
      <div
        class="form-group row"
      >
        <div
          class="p-0 col-10"
        >
          <input
            class="form-control"
            id="formSearchBox"
            placeholder="Enter search term..."
            type="text"
            value=""
          />
        </div>
        <div
          class="col-2"
        >
          <button
            class="btn btn-primary"
            type="submit"
          >
            Submit
          </button>
        </div>
      </div>
    </form>
    <div
      class="row"
    >
      <div
        class="p-0 col-12"
      >
        <div
          class="card"
          data-testid="1"
        >
          <div
            class="card-body"
          >
            <div
              class="card-title h5"
            >
              Test 1
            </div>
            <p
              class="card-text"
            >
              Salary: 100
            </p>
            <p
              class="card-text"
            >
              Equity: 0.1
            </p>
            <button
              class="btn btn-danger"
              type="button"
            >
              APPLY
            </button>
          </div>
        </div>
        <div
          class="card"
          data-testid="2"
        >
          <div
            class="card-body"
          >
            <div
              class="card-title h5"
            >
              Test 2
            </div>
            <p
              class="card-text"
            >
              Salary: 200
            </p>
            <p
              class="card-text"
            >
              Equity: 0.2
            </p>
            <button
              class="btn btn-danger"
              type="button"
            >
              APPLY
            </button>
          </div>
        </div>
        <div
          class="card"
          data-testid="3"
        >
          <div
            class="card-body"
          >
            <div
              class="card-title h5"
            >
              Test 3
            </div>
            <p
              class="card-text"
            >
              Salary: 300
            </p>
            <p
              class="card-text"
            >
              Equity: 0.3
            </p>
            <button
              class="btn btn-danger"
              type="button"
            >
              APPLY
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
